### ISSUE TITLE
Impossible to depend on azure[common]

### ISSUE THREAD

**Issue by wvxvw** [Azure/azure-sdk-for-python#3950]
Related to this ticket, however, OP didn't understand / research why the reinstall worked: https://github.com/Azure/azure-sdk-for-python/issues/1558

So, what happens:

1. `azure` package doesn't have `common` sub-package.
2. `azure-common` is an entirely separate package, so it also doesn't have `azure.common` sub-package.
3. `azure[common]` though is a special creature, which can only be installed on a full-moon night, after you sacrifice a couple of firstborns.  It is only installed if you already had `azure` installed *and then* you installed `azure-common` (well, if you are lucky).  This is because it declares a feature dependency (why oh why?) and it uses the name for this feature that just will never parse well with either `pip` or `setuptools` (because it uses dot in the name, and that's invalid).
4. Say, you got extremely lucky and you did install this elusive package (without which nothing really works, how is this an "extra", beats me...)  Well, anyways, after you install it, there is no trace of what you did!  If you try `pip freeze`, you will only see `azure` and `azure-common`, nothing about `azure[common]`.  Nothing to put in your `setup.py`.  It just vanished.

----

Bottom line, if you believe that it is possible to depend on `azure[common]`, please write an example `setup.py` which depends on it.  Else, the whole thing is really unusable.  At least through PyPA.  Only your docker distributions or your packages like DEB / RPM etc.

**Comment by wvxvw** [comment]
Further investigation revealed that some of my assumption had been wrong.  Specifically, there isn't an extra package `azure[common]` that is some asinine behavior on the side of `pip`, where when it doesn't find the package you wanted it to install, it starts installing randoms stuff.  Well... that's nice, but isn't the problem here.

The problem here is that the wheel produced from `azure-common` directory contains multiple packages:

```
M Filemode      Length  Date         Time      File
- ----------  --------  -----------  --------  -------------------------------------------
  -rw-rw-r--      1358  26-Sep-2018  21:08:40  azure/common/__init__.py
  -rw-rw-r--       327  26-Sep-2018  21:08:40  azure/common/_version.py
  -rw-rw-r--      8737  26-Sep-2018  21:08:40  azure/common/client_factory.py
  -rw-rw-r--       755  26-Sep-2018  21:08:40  azure/common/cloud.py
  -rw-rw-r--      2479  26-Sep-2018  21:08:40  azure/common/credentials.py
  -rw-rw-r--       805  26-Sep-2018  21:08:40  azure/common/exceptions.py
  -rw-rw-r--      6079  26-Sep-2018  21:08:40  azure/profiles/__init__.py
  -rw-rw-r--      3943  26-Sep-2018  21:08:40  azure/profiles/multiapiclient.py
  -rw-rw-r--         6  26-Sep-2018  21:21:54  azure_common-1.1.16.dist-info/top_level.txt
  -rw-rw-r--       110  26-Sep-2018  21:21:54  azure_common-1.1.16.dist-info/WHEEL
  -rw-rw-r--      3805  26-Sep-2018  21:21:54  azure_common-1.1.16.dist-info/METADATA
  -rw-rw-r--       997  26-Sep-2018  21:21:54  azure_common-1.1.16.dist-info/RECORD
- ----------  --------  -----------  --------  -------------------------------------------
                 29401                         12 files
```
But, `setuptools` doesn't seem to be able to install such wheels. Instead, it seems to randomly select a package from a wheel and installs that...  So, sometimes you can actually get `azure.common` installed, but that will come at the price of `azure_common` not being installed.  Or, maybe you'll get `azure.profiles`, and all else will be lost.

However, fundamentally, it's your fault.  Why did you make a wheel with more than one package in it?  This simply doesn't work.  It's impossible to depend on a package distributed in such a way.  It is probably a mistake that `pip` (kind of) installs this.

**Comment by wvxvw** [comment]
> What error do you have?

`setuptools` don't report any errors, it silently installs only a part of the wheel. Actually, the part that nobody needs.

> What pip message do you get?

`pip` has nothing to do with the problem.  I'm not using it to install the wheel.  I'm using `setuptools`.  `pip` was a confounder, that led to to believe that there exists a package, that in reality doesn't exist.  This happened because `pip` has a bug, but it is not the problem of your package distribution, it's a `pip` problem.

> The distribution package name is not related at all to the content, where you can put as many import package/modules as you want, in whatever naming you want.

This is where you are wrong.  No, you cannot put as many "import packages".  You can put only one.  `setuptools` cannot deal with more than one package.  And you **must** use setuptools, if you want to allow **library authors** to use your packages.  `pip` is unusable for library writing: you cannot use it to download your dependencies, if you are writing a library.  Effectively right now, your distribution prevents others from depending on it in a library.

**Comment by lmazuel** [comment]
I still believe I'm right with all my respect :). And that your assumption on setuptools is wrong.

Let's talk about facts. Please share a setup.py example that reproduce the issue you have, including setuptools version installed.

Thank you

**Comment by wvxvw** [comment]
You believe? Since when belief is a substitute for experiment?

There's nothing special required. *Any* `setup.py` which doesn't modify `install` command will fail to install `azure.common` package.  So, if you cannot write this, here's an example:

``` python
from setuptools import setup

setup(install_requires=['azure-common'])
```

Run `python setup.py install` then `python -c 'import azure.common'`.

**Comment by wvxvw** [comment]
Just to give you a taste of what hoops one needs to jump to depend on `azure-common`:

``` python
#!/usr/bin/env python
# -*- coding: utf-8 -*-

from setuptools import setup
from setuptools.command.install import install
from pip._internal.commands.install import InstallCommand


class Install(install):
    '''
    Customized setuptools install command which uses pip.
    '''

    def run(self, *args, **kwargs):
        install_requires = self.distribution.install_requires
        self.distribution.install_requires = []
        cmd = InstallCommand()
        options = cmd.parse_args([])[0]
        cmd.run(options, install_requires)
        super().run(*args, **kwargs)


setup(
    cmdclass={
        'install': Install,
    },
    install_requires=[
        'azure-cli-core >= 2.0.50',
        'azure-mgmt-resource >= 2.0.0',
        'azure >= 4.0.0',
        'azure-common >= 1.1.16',
    ],
    setup_requires=['wheel', 'pip'],
    zip_safe=False,
)
```

But this doesn't solve all the problems. It only solves them for the package that implements this kind of `setup()`. Further packages that will depend on this one will face the same problem, unless this package puts up additional struggle in order to execute some Python code during install (to, essentially, require that `pip` installs the packages rather than `easy_install`). But this will lead to all sorts of bad things happening potentially, beside long waiting times... `pip` version installed by `easy_install` may not be the one you want, or, you may specify not the version that your user wants... and you cannot have more than one at once. Obviously, mixing two package managers in order to install a single package is insane and will probably result in even more bugs than what I can imagine right now.

**Comment by lmazuel** [comment]
Will all my respect again, I asked you an actual way to reproduce your issue since the beginning, since I know the difference between facts and beliefs. I chose my word carefully.

Using your experiment, I don't reproduce the issue:
```python
> python .\setup.py install
running install
running bdist_egg
running egg_info
creating UNKNOWN.egg-info
writing UNKNOWN.egg-info\PKG-INFO
writing dependency_links to UNKNOWN.egg-info\dependency_links.txt
writing requirements to UNKNOWN.egg-info\requires.txt
writing top-level names to UNKNOWN.egg-info\top_level.txt
writing manifest file 'UNKNOWN.egg-info\SOURCES.txt'
reading manifest file 'UNKNOWN.egg-info\SOURCES.txt'
writing manifest file 'UNKNOWN.egg-info\SOURCES.txt'
installing library code to build\bdist.win-amd64\egg
running install_lib
warning: install_lib: 'build\lib' does not exist -- no Python modules to install

creating build
creating build\bdist.win-amd64
creating build\bdist.win-amd64\egg
creating build\bdist.win-amd64\egg\EGG-INFO
copying UNKNOWN.egg-info\PKG-INFO -> build\bdist.win-amd64\egg\EGG-INFO
copying UNKNOWN.egg-info\SOURCES.txt -> build\bdist.win-amd64\egg\EGG-INFO
copying UNKNOWN.egg-info\dependency_links.txt -> build\bdist.win-amd64\egg\EGG-INFO
copying UNKNOWN.egg-info\requires.txt -> build\bdist.win-amd64\egg\EGG-INFO
copying UNKNOWN.egg-info\top_level.txt -> build\bdist.win-amd64\egg\EGG-INFO
zip_safe flag not set; analyzing archive contents...
creating dist
creating 'dist\UNKNOWN-0.0.0-py3.7.egg' and adding 'build\bdist.win-amd64\egg' to it
removing 'build\bdist.win-amd64\egg' (and everything under it)
Processing UNKNOWN-0.0.0-py3.7.egg
Copying UNKNOWN-0.0.0-py3.7.egg to c:\users\lmazuel\documents\github\testsetup\lib\site-packages
Adding UNKNOWN 0.0.0 to easy-install.pth file

Installed c:\users\lmazuel\documents\github\testsetup\lib\site-packages\unknown-0.0.0-py3.7.egg
Processing dependencies for UNKNOWN==0.0.0
Searching for azure-common
Reading https://pypi.python.org/simple/azure-common/
C:\Users\lmazuel\Documents\GitHub\testsetup\lib\site-packages\setuptools\pep425tags.py:89: RuntimeWarning: Config variable 'Py_DEBUG' is unset, Python ABI tag may be incorrect
  warn=(impl == 'cp')):
C:\Users\lmazuel\Documents\GitHub\testsetup\lib\site-packages\setuptools\pep425tags.py:93: RuntimeWarning: Config variable 'WITH_PYMALLOC' is unset, Python ABI tag may be incorrect
  warn=(impl == 'cp')):
Downloading https://files.pythonhosted.org/packages/ac/d3/055ce7ad06459a415ff9ca210e04c6cbb51bd6564815b7c8ac34bf5a1c39/azure_common-1.1.16-py2.py3-none-any.whl#sha256=c908621a71eb4ee9fab0962e35d3c27a18f09a854d8359c2f32c15b3f4fc576e
Best match: azure-common 1.1.16
Processing azure_common-1.1.16-py2.py3-none-any.whl
Installing azure_common-1.1.16-py2.py3-none-any.whl to c:\users\lmazuel\documents\github\testsetup\lib\site-packages
writing requirements to c:\users\lmazuel\documents\github\testsetup\lib\site-packages\azure_common-1.1.16-py3.7.egg\EGG-INFO\requires.txt
Adding azure-common 1.1.16 to easy-install.pth file

Installed c:\users\lmazuel\documents\github\testsetup\lib\site-packages\azure_common-1.1.16-py3.7.egg
Finished processing dependencies for UNKNOWN==0.0.0
```

Both `python -c 'import azure.common' and python -c 'import azure.profiles' work as expected and all files.

I still have no actual factual reasons to believe there is an actual issue.

Let's ask a second opinion, @brettcannon any opinion here?

**Comment by brettcannon** [comment]
From reading this post my guess is @wvxvw is using an outdated version of setuptools that has a bug preventing it from working appropriately for them and @lmazuel  has a recent version that has an appropriate fix.

**Comment by lmazuel** [comment]
Thanks @brettcannon 
I tracked down different version of setuptools, and `azure-common` can be processed starting [17.1](https://setuptools.readthedocs.io/en/latest/history.html#id377). Before that, it fails with with:
```python
error: Setup script exited with error in azure-common setup command: Invalid environment marker: python_version<'3.0'
```

which is not the claimed issue here. I tried different version of setuptools, between this one and now, and was never able to reproduce the issue.

At best, I can update the documentation to say it requires setuptools 17.1, but this requires no other action items.

**Comment by wvxvw** [comment]
I don't care if it works on Windows.

It doesn't work on Linux with any version of `setuptools`.  What you install on Windows is an entirely different package, with an entirely different installer.

**Comment by lmazuel** [comment]
You never said Linux was your target... Again, experiment means protocol, you never describe your environment. Also, Python being cross-platform, working on Windows for things like setuptools is a good indicator of success. But yes, that's not proof, so let's test it:

```shell
lmazuel@ubuntutestvm:~$ python3 -m venv testsetup
lmazuel@ubuntutestvm:~$ . testsetup/bin/activate
(testsetup) lmazuel@ubuntutestvm:~$ vim setup.py
(testsetup) lmazuel@ubuntutestvm:~$ python3 ./setup.py install
running install
running bdist_egg
running egg_info
creating UNKNOWN.egg-info
writing UNKNOWN.egg-info/PKG-INFO
writing dependency_links to UNKNOWN.egg-info/dependency_links.txt
writing requirements to UNKNOWN.egg-info/requires.txt
writing top-level names to UNKNOWN.egg-info/top_level.txt
writing manifest file 'UNKNOWN.egg-info/SOURCES.txt'
reading manifest file 'UNKNOWN.egg-info/SOURCES.txt'
writing manifest file 'UNKNOWN.egg-info/SOURCES.txt'
installing library code to build/bdist.linux-x86_64/egg
running install_lib
warning: install_lib: 'build/lib' does not exist -- no Python modules to install

creating build
creating build/bdist.linux-x86_64
creating build/bdist.linux-x86_64/egg
creating build/bdist.linux-x86_64/egg/EGG-INFO
copying UNKNOWN.egg-info/PKG-INFO -> build/bdist.linux-x86_64/egg/EGG-INFO
copying UNKNOWN.egg-info/SOURCES.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying UNKNOWN.egg-info/dependency_links.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying UNKNOWN.egg-info/requires.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
copying UNKNOWN.egg-info/top_level.txt -> build/bdist.linux-x86_64/egg/EGG-INFO
zip_safe flag not set; analyzing archive contents...
creating dist
creating 'dist/UNKNOWN-0.0.0-py3.6.egg' and adding 'build/bdist.linux-x86_64/egg' to it
removing 'build/bdist.linux-x86_64/egg' (and everything under it)
Processing UNKNOWN-0.0.0-py3.6.egg
Copying UNKNOWN-0.0.0-py3.6.egg to /home/lmazuel/testsetup/lib/python3.6/site-packages
Adding UNKNOWN 0.0.0 to easy-install.pth file

Installed /home/lmazuel/testsetup/lib/python3.6/site-packages/UNKNOWN-0.0.0-py3.6.egg
Processing dependencies for UNKNOWN==0.0.0
Searching for azure-common
Reading https://pypi.python.org/simple/azure-common/
Downloading https://files.pythonhosted.org/packages/ac/d3/055ce7ad06459a415ff9ca210e04c6cbb51bd6564815b7c8ac34bf5a1c39/azure_common-1.1.16-py2.py3-none-any.whl#sha256=c908621a71eb4ee9fab0962e35d3c27a18f09a854d8359c2f32c15b3f4fc576e
Best match: azure-common 1.1.16
Processing azure_common-1.1.16-py2.py3-none-any.whl
Installing azure_common-1.1.16-py2.py3-none-any.whl to /home/lmazuel/testsetup/lib/python3.6/site-packages
writing requirements to /home/lmazuel/testsetup/lib/python3.6/site-packages/azure_common-1.1.16-py3.6.egg/EGG-INFO/requires.txt
Adding azure-common 1.1.16 to easy-install.pth file

Installed /home/lmazuel/testsetup/lib/python3.6/site-packages/azure_common-1.1.16-py3.6.egg
Finished processing dependencies for UNKNOWN==0.0.0
(testsetup) lmazuel@ubuntutestvm:~$ python3 -c 'import azure.common'
(testsetup) lmazuel@ubuntutestvm:~$ python3 -c 'import azure.profiles'
(testsetup) lmazuel@ubuntutestvm:~$ python3 -c 'import setuptools; print(setuptools.__version__)'
39.0.1
```

Ubuntu bionic. Your problem can't be reproduced.
